### PR TITLE
ARROW-11158: [Julia] Implement Decimal256 support for Julia

### PR DESCRIPTION
**Pull request by quinnj**
This PR also includes a few other bugfixes since the original Julia code
donation. The details of the release process of the code here and
transition from the JuliaData/Arrow.jl repository are still being ironed
out, so there has been development in both places and this synchronizes
the two.

cc: @nealrichardson

### CODE REVIEW

File: julia/Arrow/Project.toml [hunk 1]
```diff
@@ -1,9 +1,10 @@
 name = "Arrow"
 uuid = "69666777-d1a9-59fb-9406-91d4454c9d45"
 authors = ["quinnj <quinn.jacobd@gmail.com>"]
-version = "0.3.0"
+version = "1.1.0"
```
Comment (nealrichardson) [inline julia/Arrow/Project.toml:4]: Apache Arrow libraries are currently at 2.0.0 and are about to release 3.0.0; should this version number track that as well?

File: julia/Arrow/README.md [hunk 2]
```diff
@@ -1,13 +1,38 @@
...
 
+## Installation
+
+The package can be installed by typing in the following in a Julia REPL:
+
+```julia
+julia> using Pkg; Pkg.add(url="https://github.com/apache/arrow", subdir="julia/Arrow.jl")
+```
+
+or to use the non-official-apache code that will sometimes include bugfix patches between apache releases, you can do:
```
Comment (nealrichardson) [inline julia/Arrow/README.md:24]: I believe "Apache" should be capitalized throughout

File: julia/Arrow/README.md [hunk 3]
```diff
@@ -1,13 +1,38 @@
...
+
+```julia
+julia> using Pkg; Pkg.add("Arrow")
+```
+
+## Difference between this code and the JuliaData/Arrow.jl repository
+
+This code is officially part of the apache/arrow repository and as such follows the regulated release cadence of the entire project, following standard community
+voting protocols. The JuliaData/Arrow.jl repository can be viewed as a sort of "dev" or "latest" branch of this code that may release more frequently, but without following
+official apache release guidelines. The two repositories are synced, however, so any bugfix patches in JuliaData will be upstreamed to apache/arrow for each release.
```
Comment (nealrichardson) [inline julia/Arrow/README.md:34]: I would encourage you to consider inverting this workflow: do your issue tracking and bugfixes in apache/arrow, and if you need to cut a bugfix release, sync from here to JuliaData/Arrow.jl.

File: julia/Arrow/docs/src/manual.md [hunk 4]
```diff
@@ -55,10 +55,50 @@ In the arrow data format, specific logical types are supported, a list of which
...
 * `Char` and `Symbol` Julia types are mapped to arrow string types, with additional metadata of the original Julia type; this allows deserializing directly to `Char` and `Symbol` in Julia, while other language implementations will see these columns as just strings
-* `Decimal128` has no corresponding builtin Julia type, so it's deserialized using a compatible type definition in Arrow.jl itself: `Arrow.Decimal`
+* `Decimal128` and `Decimal256` have no corresponding builtin Julia types, so they're deserialized using a compatible type definition in Arrow.jl itself: `Arrow.Decimal`
 
 Note that when `convert=false` is passed, data will be returned in Arrow.jl-defined types that exactly match the arrow definitions of those types; the authoritative source for how each type represents its data can be found in the arrow [`Schema.fbs`](https://github.com/apache/arrow/blob/master/format/Schema.fbs) file.
 
+#### Custom types
+
+To support writing your custom Julia struct, Arrow.jl utilizes the format's mechanism for "extension types" by storing
+the Julia type name in the field metadata. To "hook in" to this machinery, custom types can just call
```
Comment (nealrichardson) [inline julia/Arrow/docs/src/manual.md:65]: FWIW, both Pandas and R store custom metadata that only they recognize in the Schema's key-value metadata, not as extension types. By our understanding, extension types make most sense when it's a type definition that is to be supported in other languages/implementations of Arrow. That's often not the case when you're looking just to make sure you have high fidelity when write/reading data to Feather/Parquet/etc.

That's not to say that it is wrong to do what you're doing here: just sharing how we in python/R have worked through similar issues.

**Review by nealrichardson — COMMENTED** [review]
A few notes, for what it's worth.

Since the Julia GitHub Actions are still blocked on apache/arrow, can you please look at the actions run on your fork and attach a link to a passing build there? They're not blocked from running on forks.

File: julia/Arrow/Project.toml (hunk 1 above)
Comment (quinnj) [reply julia/Arrow/Project.toml:4]: This version number tracks the currently registered package in the Julia registry; for the Apache version, I plan on updating the instructions in the README with the `Pkg` command to install the library release version according to the apache/arrow release tag (i.e. `apache-arrow-2.0.0`); that will ensure people can install the exact, official Apache release code.

File: julia/Arrow/README.md (hunk 3 above)
Comment (quinnj) [reply julia/Arrow/README.md:34]: Well, we currently do a bugfix release after every bugfix commit in JuliaData. There are also the complications with apache/arrow CI right now that make it pretty difficult to develop against. Another point is that Julia packages have historically been their own github repos, and so users naturally go to the github repo to file issues or make PRs. There's much lower friction, institutionally, to do that at JuliaData and then ensure changes get moved up here.

I do think it's something we can transition over time though, with effort. I think the Julia ecosystem will become easier to allow a package to live in a monorepo and dev efforts can be encouraged here.

**Comment by nealrichardson** [comment]
@quinnj if you rebase (or push additional changes), the Julia CI should be re-enabled.

**Pull request closed without merging**